pennylane_forest/qvm.py

```diff
@@ -165,80 +144,66 @@ def pre_measure(self):
-
-            for w, b in zip(eigvals, itertools.product([0, 1], repeat=len(wires))):
-                samples = np.where(np.all(res == b, axis=1), w, samples)
-
-            return samples
-
-        return 1 - 2 * self.state[wires[0]]
+    def apply_parametric_program(self, operations, **kwargs):
+        # pylint: disable=attribute-defined-outside-init
+        rotations = kwargs.get("rotations", [])
+
+        # Storing the active wires
+        self._active_wires = ForestDevice.active_wires(operations + rotations)
+
+        # Apply the circuit operations
+        for i, operation in enumerate(operations):
+            # number of wires on device
+            wires = self.remap_wires(operation.wires)
+
+            if i > 0 and operation.name in ("QubitStateVector", "BasisState"):
+                raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
+                                  "on a {} device.".format(operation.name, self.short_name))
+
+            # Prepare for parametric compilation
+            par = []
+            for param in operation.params:
+                if isinstance(param, Variable):
+                    # Using the idx for each Variable instance
+                    parameter_string = "theta" + str(param.idx)
+                    if parameter_string not in self._parameter_map:
+                        current_ref = self.prog.declare(parameter_string, "REAL")
+                        self._parameter_reference_map[parameter_string] = current_ref
+
+                    self._parameter_map[parameter_string] = [param.val]
+
+                    # Appending the parameter reference
+                    par.append(self._parameter_reference_map[parameter_string])
+                else:
+                    par.append(param)
 
-    def probabilities(self, wires):
-        """Returns the (marginal) probabilities of the quantum state.
+            self.prog += self._operation_map[operation.name](*par, *wires)
 
-        Args:
-            wires (Sequence[int]): sequence of wires to return
-                marginal probabilities for. Wires not provided
-                are traced out of the system.
+        self.apply_rotations(rotations)
 
-        Returns:
-            array: array of shape ``[2**len(wires)]`` containing
-            the probabilities of each computational basis state
-        """
-        # create an array of size [2^len(wires), 2] to store
-        # the resulting probability of each computational basis state
-        probs = np.zeros([2 ** len(wires), 2])
-        probs[:, 0] = np.arange(2 ** len(wires))
+    def generate_samples(self):
+        if "pyqvm" in self.qc.name:
+            self._samples = self.qc.run(self.prog, memory_map=self._parameter_map)
+        else:
+            # No hash provided or parametric compilation was set to False
+            # Will compile the program
+            if self.circuit_hash is None or not self.parametric_compilation:
```

Comment:
Where is `circuit_hash` getting set?

Reply:
Hi, yes, that's a completely valid question, as based on only this PR it comes as something out of context.

Following up on the previous PR #38 , now ``ForestDevice`` inherits from ``QubitDevice``. The PennyLane type of flow of logic involves ``evaluating`` a quantum circuit by calling on the ``execute`` method of the device (such is ``QVMDevice``).

Each instance of the ``ForestDevice`` (now ``QVMDevice`` in our case) will currently inherit the same ``execute`` method [as seen here](https://github.com/XanaduAI/pennylane/blob/a1f22b10ee968ade10488c6c42a7b153553cf628/pennylane/_qubit_device.py#L123) (which is not redefined in any of the subclasses).

After some checks, the first statement of this method is 
``self._circuit_hash = circuit.hash``.

Here, ``circuit.hash`` is a hash created from the circuit for the python session [as seen here](https://github.com/XanaduAI/pennylane/blob/a1f22b10ee968ade10488c6c42a7b153553cf628/pennylane/circuit_graph.py#L187) and [here](https://github.com/XanaduAI/pennylane/blob/a1f22b10ee968ade10488c6c42a7b153553cf628/pennylane/circuit_graph.py#L140). Identical circuits will produce identical hashes. When it comes to comparing if parametric gates were called with the arguments, [numeric parameters are checked by value](https://github.com/XanaduAI/pennylane/blob/a1f22b10ee968ade10488c6c42a7b153553cf628/pennylane/circuit_graph.py#L165), [symbolic parameters are checked by reference](https://github.com/XanaduAI/pennylane/blob/a1f22b10ee968ade10488c6c42a7b153553cf628/pennylane/circuit_graph.py#L158). (In the implementation itself, a string is being built).

Reply:
Just to sketch the flow of logic in ``execute``:
1) The circuit hash is set
2) The circuit operations and rotations into the measurement basis are applied (in our case this means building up the PyQuil ``Program``)
3) Samples are generated in the computational basis (in our case this involves compiling (if applicable) and running the ``Program``)
4) Statistics are extracted based on the samples obtained (e.g. returning expectation values, variances, etc.)